contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -255,6 +263,78 @@ pub fn set_chain_config(
     }
 }
 
+/// Calculates the destination on token transfer amount.
+///
+/// The amount is calculated based on the token decimals on the source and destination chains.
+/// The calculation is done as following:
+/// 1) `destination_amount` = `source_amount` * 10 ^ (`destination_chain_decimals` - `source_chain_decimals`)
+/// 3) If new_amount is greater than the destination chain's `max_uint`, the translation
+/// fails.
+/// 4) If new_amount is zero, the translation fails.
+fn destination_amount(
+    storage: &dyn Storage,
+    source_chain: &ChainNameRaw,
+    destination_chain: &ChainNameRaw,
+    token_id: TokenId,
+    source_amount: nonempty::Uint256,
+) -> Result<nonempty::Uint256, Error> {
+    let source_token = try_load_token_instance(storage, source_chain.clone(), token_id)?;
+    let destination_token = try_load_token_instance(storage, destination_chain.clone(), token_id)?;
+    let (source_decimals, destination_decimals) =
+        match (source_token.decimals, destination_token.decimals) {
+            (Some(source_decimals), Some(destination_decimals)) => {
+                (source_decimals, destination_decimals)
+            }
+            (None, None) => return Ok(source_amount),
+            _ => unreachable!(
+                "decimals are either set in both the source and destination, or set in neither"
+            ), // This should never happen
+        };
+    let destination_max_uint = state::load_chain_config(storage, destination_chain)
+        .change_context(Error::State)?
+        .max_uint;
+
+    // It's intentionally written in this way since the end result may still be fine even if
+    //     1) amount * (10 ^ (dest_chain_decimals)) overflows
+    //     2) amount / (10 ^ (src_chain_decimals)) is zero
+    let scaling_factor = Uint256::from_u128(10)
+        .checked_pow(source_decimals.abs_diff(destination_decimals).into())
+        .change_context_lazy(|| Error::InvalidTransferAmount {
+            source_chain: source_chain.to_owned(),
+            destination_chain: destination_chain.to_owned(),
+            amount: source_amount,
+        })?;
+    let destination_amount = if source_decimals > destination_decimals {
+        source_amount
+            .checked_div(scaling_factor)
+            .expect("scaling_factor must be non-zero")
+    } else {
+        source_amount
+            .checked_mul(scaling_factor)
+            .change_context_lazy(|| Error::InvalidTransferAmount {
+                source_chain: source_chain.to_owned(),
+                destination_chain: destination_chain.to_owned(),
+                amount: source_amount,
+            })?
+    };
+
+    if destination_amount.gt(&destination_max_uint) {
+        bail!(Error::InvalidTransferAmount {
+            source_chain: source_chain.to_owned(),
+            destination_chain: destination_chain.to_owned(),
+            amount: source_amount,
+        })
```

Comment:
Feel like this should be something like `OverflowingTransfer` and also include the destination_max_uint in the error

Reply:
I'd prefer keeping this simple and generic since all the other cases also suggest some very specific number issue.